services/placement/algo/sharded.go

```diff
@@ -157,10 +158,13 @@ func (a rackAwarePlacementAlgorithm) ReplaceInstance(
 }
 
 func (a rackAwarePlacementAlgorithm) addInstance(p services.ServicePlacement, addingInstance services.PlacementInstance) (services.ServicePlacement, error) {
-	if _, exist := p.Instance(addingInstance.ID()); exist {
-		return nil, errAddingInstanceAlreadyExist
-
+	if instance, exist := p.Instance(addingInstance.ID()); exist {
```

Comment:
Can we do the following instead so `movingLeavingShardsBack` doesn't need to return the instance?
```
if _, exists := p.Instance(addingInstance.ID()); exist {
      if !placement.IsInstanceLeaving(addingInstance) {
          return nil, errAddingInstanceAlreadyExist
      }
      p = a.moveLeavingShardsBack(p, addingInstance)
}
```